### DISCUSSION TITLE
POST Request Fails on the destination?

### DISCUSSION THREAD

**Discussion by brenner23** [arendst/Tasmota#19859] [Q&A]
Hello ive got some lines to send a POST
```
def GetData()
	var cl = webclient()
	cl.set_useragent('ESP32HTTPClientTasmotaESP32')        
	cl.add_header('Content-Type','application/x-www-form-urlencoded')

	# Connect to my Server(s) like Rented server 
        #cl.begin('http://192.168.0.79/Test-POST.php')    <<< Mein PC Server in my Lan
	
        #Connect to Test Server on the internet
        cl.begin('https://httpbin.org/post') <<< Testpage to check  POST Resquest

	print(cl.POST('API_KEY=SomeContent')) <<< here the Payload
	print(cl.get_string()) 
	cl.close()
end
GetData()
```

php file content:
```
<?php
echo "{Debug: API_KEY --> ".$_POST[API_KEY]."}";
?>
```

My outputs
https://httpbin.org/post:
...
  "data": "API_KEY=SomeContent", 
...
thats OK 
my Server and same output from my Rentet server where i have only access to FTP
http://192.168.0.79/Test-POST.php:
{Debug: API_KEY --> }

if i check with Firefox Addon Http request maker , than looks ok if i send a Post with Content 
output is {Debug: API_KEY -->SomePayload }

and on Apache logs i see 200 .... so Webside reqest ok but no Payload $_:POST['API_KEY ']... 

my question:
is there someversions  on lybary  where start POST and my server wants some more Requierd info  as safety?
Doi can (must) Add any in rhe Berry lines ?

ist some different on POST from Tasmota to Normaly Post ?
My server is installed with Debian and install apache php...... so nothing fancy :-)
You could try to recreate it and explain to me where the error is .

**Comment by barbudor** [comment]
Very difficult to follow your explanations

Does a curl works?

**Reply by brenner23** [reply]
hello 

crurl from Tasmota frommy pc ?

btw 
{Debug: API_KEY --> }  <<< this output comes from webside  

Befor the lines was Post i was use  GET and that works...
reason is i donte like  http://my side.php?MyPassword=ljlkjlk3jlk213&some=other secrets

**Reply by s-hadinger** [reply]
You should enable detailed logging on your server to understand what payload is received

**Reply by barbudor** [reply]
Curl from a PC of course, not the same as the debian one running the apache

Could you make some effort in your typing? Your messages are barely readable. Thanks

**Reply by brenner23** [reply]
Hello,
here my serverlogs

access.log
192.168.0.114 - - [28/Oct/2023:01:24:55 +0200] "POST /Test-POST.php HTTP/1.1" 200 576 "-" "ESP32HTTPClientTasmotaESP32"
192.168.0.114 - - [28/Oct/2023:01:25:15 +0200] "POST /Test-POST.php HTTP/1.1" 200 576 "-" "ESP32HTTPClientTasmotaESP32"
192.168.0.114 - - [28/Oct/2023:01:27:22 +0200] "POST /Test-POST.php HTTP/1.1" 200 576 "-" "ESP32HTTPClientTasmotaESP32"

and  error logs  why empty  post 

[Sat Oct 28 13:45:11.274046 2023] [php7:warn] [pid 160826] [client 192.168.0.114:61472] PHP Warning:  Use of undefined constant API_KEY - assumed 'API_KEY' (this will throw an Error in a future version of PHP) in /var/www/html/Test-POST.php on line 6

**Reply by brenner23** [reply]
so  thats  my logs i can get , or  can i get logs from php7 ?

**Reply by barbudor** [reply]
First you have a syntax error in your PHP code as API_KEY should be 'API_KEY'

Then your POST payload should be a JSON string and add the header Content-type: application/json

**Reply by brenner23** [reply]
i copy some into file ... 

phpfile i have add some

```
<?php
echo "<pre>"; print_r($_POST) ;  echo "</pre>";
echo "<hr>";
echo "{Debug: API_KEY --> ".$_POST['API_KEY']."}";
?>
```


Welcome to the Berry Scripting console. Check the documentation.
```
def GetData()
	var cl = webclient()
	cl.set_useragent('ESP32HTTPClientTasmotaESP32')        
	cl.add_header('Content-Type','application/x-www-form-urlencoded')
	#cl.begin('http://192.168.0.79/Test-POST.php')
	cl.begin('https://httpbin.org/post')
	print(cl.POST('API_KEY=raspberry'))
	print(cl.get_string())
	cl.close()
end
GetData()
```
my Server has no JSON output  only this 
```
<pre>Array
(
)
</pre><hr><br />
<b>Notice</b>:  Undefined index: API_KEY in <b>/var/www/html/Test-POST.php</b> on line <b>6</b><br />
{Debug: API_KEY --> }

```


here the output fromTestside

```
def GetData()
	var cl = webclient()
	cl.set_useragent('ESP32HTTPClientTasmotaESP32')        
	cl.add_header('Content-Type','application/x-www-form-urlencoded')
	#cl.begin('http://192.168.0.79/Test-POST.php')
	cl.begin('https://httpbin.org/post') <<<<--- Test side
	print(cl.POST('API_KEY=raspberry'))
	print(cl.get_string())
	cl.close()
end
GetData()
```


```


{
  "args": {}, 
  "data": "API_KEY=raspberry", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept-Encoding": "identity;q=1,chunked;q=0.1,*;q=0", 
    "Content-Length": "17", 
    "Host": "httpbin.org", 
    "User-Agent": "ESP32HTTPClientTasmotaESP32", 
    "X-Amzn-Trace-Id": "Root=1-653cf9c4-0af3191b5aa5638b17a3894c"
  }, 
  "json": null, 
  "origin": "1xxxxxxxx", 
  "url": "https://httpbin.org/post"
}
```

**Reply by brenner23** [reply]
and this line tells me that  tespage works

`"data": "API_KEY=raspberry", `

**Reply by sfromis** [reply]
Looks like the same slow tooth-pulling as on Discord, slowly leading to there being no issue at all.
https://discord.com/channels/479389167382691863/826169659811168276/1167785958750552105

**Reply by brenner23** [reply]
problen is fixed now .... ty
